### PR TITLE
ATO-661: Add method to generate storage access token for Auth to send to IPV

### PR DESCRIPTION
**Pull request by billrdunn**
## What

Add method to generate a signed storage access token. 

Authentication will send this storage token to IPV as a claim (named https://vocab.account.gov.uk/v1/storageAccessToken) in the JAR when identity re-verification is required. More details [here](https://govukverify.atlassian.net/wiki/spaces/Architecture/pages/4188799112/Storage+access+token).

This is required as part of the MFA reset initiative.

## How to review

1. Code review only
2. Confirm that sonarcloud analysis can be dismissed because:
  - ConfigurationService tests are not required
  - Code duplication is between authentication shared and orchestration shared which is expected

## Checklist

- [x] Impact on orch and auth mutual dependencies has been checked.

### CODE REVIEW

File: shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java [hunk 1]
```diff
@@ -514,4 +514,26 @@ public long getAccountInterventionServiceCallTimeout() {
     public String getAccountInterventionsErrorMetricName() {
         return System.getenv().getOrDefault("ACCOUNT_INTERVENTIONS_ERROR_METRIC_NAME", "");
     }
+
+    public String getIPVAudience() {
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:518]: Could you leave a comment on the ticket that uses this to pass this variable please?

Comment (billrdunn) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/ConfigurationService.java:518]: [Done](https://govukverify.atlassian.net/browse/ATO-660?focusedCommentId=152305)

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java [hunk 2]
```diff
@@ -278,14 +278,43 @@ private SignedJWT generateIDToken(
...
+        LOG.info("Storage token being created with JWTID: {}", jwtID);
+
+        List<String> aud =
+                List.of(
+                        configService.getCredentialStoreURI().toString(),
+                        configService.getIPVAudience());
+
+        JWTClaimsSet.Builder claimSetBuilder =
+                new JWTClaimsSet.Builder()
+                        .issuer(configService.getFrontendBaseUrl())
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:304]: Let's go with this for now, but I wonder if this should be the auth frontend API domain instead? Let's start a converstation about this with Auth and VC Storage

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java [hunk 3]
```diff
@@ -278,14 +278,43 @@ private SignedJWT generateIDToken(
...
+                        .audience(aud)
+                        .expirationTime(expiryDate)
+                        .issueTime(NowHelper.now())
+                        .subject(internalPairwiseSubject.getValue())
+                        .jwtID(jwtID);
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:309]: Think we're missing the `reverification` scope here?

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java (hunk 2 above)
Comment (ethanmills) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:304]: This one doesn't block merge

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java (hunk 3 above)
Comment (ethanmills) [reply shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:309]: This one does

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java [hunk 4]
```diff
@@ -352,13 +384,29 @@ private RefreshToken generateAndStoreRefreshToken(
         return refreshToken;
     }
 
-    public SignedJWT generateSignedJWT(
+    public SignedJWT generateSignedJwtUsingExternalKey(
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:387]: Can this be private?

File: shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java [hunk 5]
```diff
@@ -352,13 +384,29 @@ private RefreshToken generateAndStoreRefreshToken(
...
+    private SignedJWT generateSignedJwtUsingStorageKey(
+            JWTClaimsSet claimsSet, Optional<String> type) {
+        return generateSignedJWT(
+                claimsSet,
+                type,
+                JWSAlgorithm.ES256,
+                configService.getStorageTokenSigningKeyAlias());
+    }
+
+    public SignedJWT generateSignedJWT(
```
Comment (ethanmills) [inline shared/src/main/java/uk/gov/di/authentication/shared/services/TokenService.java:405]: Same for this, can it be private?

**Review by ethanmills — DISMISSED** [review]

**Comment by ethanmills** [comment]
I've noticed that this make two calls to KMS, one to fetch the key ID from the alias, then another to sign. How much do we gain by using the alias instead of the key ID directly? Is it worth that extra call? This applies every time we sign something, so it's potentially quite a lot of overhead

**Review by ethanmills — APPROVED** [review]

**Pull request merged**